### PR TITLE
binary DNS record parsing (CBOR)

### PR DESCRIPTION
**Pull request by robertmin1**
binary DNS record parsing (CBOR)

### CODE REVIEW

File: ncdomain/convert.go [hunk 1]
```diff
@@ -353,10 +355,23 @@ func ParseValue(name, jsonValue string, resolve ResolveFunc, errFunc ErrorFunc)
...
-		errFunc.add(err)
-		return
+	errJSON := json.Unmarshal([]byte(jsonValue), &rv)
+	if errJSON != nil {
+		valueBytes, errHex := hex.DecodeString(jsonValue)
+		if errHex != nil {
+			errFunc.add(errJSON)
+			errFunc.add(fmt.Errorf("failed to decode hex (after JSON decoding failed): %w", errHex))
+			return
+		}
```
Comment (JeremyRand) [inline ncdomain/convert.go:365]: What's the intent of decoding as hex? Namecoin values are never hex-encoded in the blockchain by convention. Is there some btcd-related API that's returning a hex string here?

Comment (robertmin1) [reply ncdomain/convert.go:365]: The CBOR data is encoded in hex format.

Comment (JeremyRand) [reply ncdomain/convert.go:365]: Not sure I follow -- where is it hex-encoded? It's not hex-encoded in the blockchain (that would double the storage costs for no reason).

Comment (robertmin1) [reply ncdomain/convert.go:365]: I thought when submitting `cbor data` you convert it to `hex` when pass it as a value from a name

**Comment by robertmin1** [comment]
Domain used for testing `cbortestingdomain.bit`

**Comment by JeremyRand** [comment]
> Made the relevant changes, this PR depends on [namecoin/ncrpcclient#6](https://github.com/namecoin/ncrpcclient/pull/6) (Following your suggestion to patch ncrpcclient) Would it be more appropriate to patch it in ncdns [instead](https://github.com/namecoin/ncdns/blob/master/namecoin/namecoin.go#L30)?

Patching ncrpcclient is correct.

File: ncdomain/convert.go [hunk 2]
```diff
@@ -1083,3 +1092,55 @@ func (v *Value) moveEmptyMapItems() {
 		}
 	}
 }
+
+// makeJSONSafe walks an interface to ensure all maps use string keys so that
```
Comment (JeremyRand) [inline ncdomain/convert.go:1096]: Can you include an attribution to exactly where this code came from? The GitHub issue link doesn't make it clear.

**Comment by JeremyRand** [comment]
Code review of b7a8f2053e558f2fdfab30bcc86ecb0fa6877274 looks good to me, other than the attribution issue I noted. Haven't actually tested it yet.

**Comment by robertmin1** [comment]
Updated the code

**Comment by JeremyRand** [comment]
ACK 5cb932251295fdb0062a6c7f6ac3fd3fb4662aaa

**Pull request merged**